From c437a3349b39fd54ae60df9e45e602c56f764fd3 Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@nxp.com>
Date: Tue, 26 Jun 2018 17:48:22 +0800
Subject: [PATCH 4117/5242] MLK-17921-1 staging: typec: tcpci: improve
 properties parse

commit  0d2d088b6253a5fc02153c0188bedfcb61da525e from
https://source.codeaurora.org/external/imx/linux-imx.git

Current typec and PD properties parse can't work well for sink or source
only port, this patch is to improve it.

Acked-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Li Jun <jun.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/staging/typec/tcpci.c |   25 +++++++++++++++----------
 1 file changed, 15 insertions(+), 10 deletions(-)

diff --git a/drivers/staging/typec/tcpci.c b/drivers/staging/typec/tcpci.c
index 2517394..c7a6f3d 100644
--- a/drivers/staging/typec/tcpci.c
+++ b/drivers/staging/typec/tcpci.c
@@ -694,18 +694,13 @@ static int tcpci_parse_config(struct tcpci *tcpci)
 		return -EINVAL;
 	}
 
-	/* Get the default-role */
-	tcfg->default_role = typec_get_power_role(tcpci->dev);
-	if (tcfg->default_role == TYPEC_ROLE_UNKNOWN) {
-		dev_err(tcpci->dev, "typec power role is NOT correct!\n");
-		return -EINVAL;
-	}
+	if (tcfg->type == TYPEC_PORT_UFP)
+		goto sink;
 
 	/* Check source pdo array size */
 	tcfg->nr_src_pdo = device_property_read_u32_array(tcpci->dev,
 						"src-pdos", NULL, 0);
-	if (tcfg->nr_src_pdo <= 0 && (tcfg->type == TYPEC_PORT_DRP ||
-					tcfg->type == TYPEC_PORT_DFP)) {
+	if (tcfg->nr_src_pdo <= 0) {
 		dev_err(tcpci->dev, "typec source pdo is missing!\n");
 		return -EINVAL;
 	}
@@ -724,6 +719,16 @@ static int tcpci_parse_config(struct tcpci *tcpci)
 		return -EINVAL;
 	}
 
+	if (tcfg->type == TYPEC_PORT_DFP)
+		return 0;
+
+	/* Get the default-role */
+	tcfg->default_role = typec_get_power_role(tcpci->dev);
+	if (tcfg->default_role == TYPEC_ROLE_UNKNOWN) {
+		dev_err(tcpci->dev, "typec power role is NOT correct!\n");
+		return -EINVAL;
+	}
+
 	/*
 	 * In case DRP only for data role, power role is source only
 	 * we can use this property to disable power sink.
@@ -757,11 +762,11 @@ static int tcpci_parse_config(struct tcpci *tcpci)
 		return 0;
 	}
 
+sink:
 	/* Check the num of snk pdo */
 	tcfg->nr_snk_pdo = device_property_read_u32_array(tcpci->dev,
 						"snk-pdos", NULL, 0);
-	if (tcfg->nr_snk_pdo <= 0 && (tcfg->type == TYPEC_PORT_DRP ||
-					tcfg->type == TYPEC_PORT_UFP)) {
+	if (tcfg->nr_snk_pdo <= 0) {
 		dev_err(tcpci->dev, "typec sink pdo is missing!\n");
 		return -EINVAL;
 	}
-- 
1.7.9.5

